controllers/vspherecluster_reconciler.go

```diff
@@ -127,6 +127,13 @@ func (r clusterReconciler) Reconcile(_ goctx.Context, req ctrl.Request) (_ ctrl.
 		return r.reconcileDelete(clusterContext)
 	}
 
+	// If the VSphereCluster doesn't have our finalizer, add it.
+	// Requeue immediately after adding finalizer to avoid the race condition between init and delete
+	if !ctrlutil.ContainsFinalizer(vsphereCluster, infrav1.ClusterFinalizer) {
```

Comment:
Nit/Optional: Maybe we should move this above L121 (where we call `setOwnerRefsOnVsphereMachines(...)`), because `setOwnerRefsOnVsphereMachines` may patch vspheremachines and on errors we won't have yet set the finalizer. (Note: we would have to add vsphereCluster.DeletionTimestamp.IsZero() to the if, if we are above the deletion part).

However, I think we are fine as is too.

Reply:
I think it won't make much of a difference. The main goal with the finalizer is that we set it before the current reconciler creates anything for the current object

Reply:
Maybe kepp it here as it's slightly simpler